doc/decisions/0_drafts/errors_core.md

```diff
@@ -0,0 +1,292 @@
+  This can be distinguished after the fact, by checking whether the `Key` has a read-only name with `keyIsLocked`.
+
+- Callers normally don't care about the exact error, unless they can do something about it.
+  This could be:
+
+  - reporting the error to a user
+  - switching to a different code path
+  - (during debugging) changing the code
+
+- Functions can be written in a such a way that they detect and report all errors without permanent effects.
+
+- The functions that cannot use a `Key * errorKey`, are side-effect-free.
+  Therefore, the only possible cause of errors are invalid arguments.
+
+## Considered Alternatives
+
+### Extra argument
+
+All functions that report errors take a `int * error` argument to which errors can be written.
+
+This makes the API much more convoluted to use.
+
+A benefit would be the return value remains fully usable.
+
+### Thread Local Storage
+
+Use a field in thread local storage (TLS) to store errors similar to `errno`.
+
+This leaves the API design entirely open.
+
+TLS was only standardized in C11, while Elektra currently only requires C99.
+
+This has all the same issues as `errno`: needs to be cleared to detect errors, can only store one error (per thread), etc.
+
+### Redesign API
+
+Redesign the API, so that every API function only has maximal one error case next to memory errors.
+
+- `nameNew(name) -> name_object` only has the error case of invalid name
+- `keyNew(name_object)` won't have error cases anymore
+- `keySetName(name_object)` only has the error case of read-only names
+
+This can be very limiting and in some cases makes it very hard to find out what the exact problem was.
+Reducing a function to a single error value, makes sense if the errors are easy to distinguish.
+But for more complex preconditions, this can be annoying and inefficient to use.
+While the above example can be solved, by simple checking with `keyIsLocked`, there could less straightforward cases.
+For example, a function `foo` that takes two keynames, both of which must be valid.
+To find out which of the names was invalid, the caller would have to call something like `elektraCheckName(name)` on at least one name.
+That's a non-trivial check that is not cheap in terms of runtime.
+More importantly, it is a check that `foo` already did, and the result could just be propagated to the caller.
+
+### Require caller to check preconditions
+
+If caller are interested in the specific errors, they need to do extra checks, e.g., before calling `keySetName`, they call `elektraCheckName(name)` or `keyIsLocked`.
+
+This is not so intuitive and is against our principle to make it hard to use the API wrong.
+
+It's also unclear, what the function would do, if the caller did not check the preconditions.
+
+### Pragmatic solution
+
+We assume the functions in question are side-effect-free and the only possible cause of errors are invalid arguments.
```

Comment:
This is a constraints for the core, please put it to constraints.

Reply:
It actually already is an assumption in the "Assumptions" section above. I can't make it a constraint, because it's only a constraint for `libelektra-core` not necessarily all APIs that don't (want to) use `Key * errorKey`.

I just mentioned it here as a reminder of the assumption and so the following conclusions are easier to understand, i.e., instead of saying "Because of our assumptions above" and I repeat the assumptions.